pkg/cmd/cli/cmd/rsync.go

```diff
@@ -0,0 +1,411 @@
+	Destination    string
+	DestinationDir string
+	RshCommand     string
+	UseTar         bool
+	Quiet          bool
+	Delete         bool
+
+	Out    io.Writer
+	ErrOut io.Writer
+
+	LocalExecutor  executor
+	RemoteExecutor executor
+	Tar            tar.Tar
+	PodClient      kclient.PodInterface
+}
+
+// executor executes commands
+type executor interface {
+	Execute(command []string, in io.Reader, out, err io.Writer) error
+}
+
+// NewCmdRsync creates a new sync command
+func NewCmdRsync(name, parent string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	tarHelper := tar.New()
+	tarHelper.SetExclusionPattern(nil)
+	o := RsyncOptions{
+		Out:           out,
+		ErrOut:        errOut,
+		LocalExecutor: &defaultLocalExecutor{},
+		Tar:           tarHelper,
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s SOURCE_DIR POD:DESTINATION_DIR", name),
+		Short:   "Copy local files to a pod",
+		Long:    rsyncLong,
+		Example: fmt.Sprintf(rsyncExample, parent+" "+name),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(f, c, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.RunRsync())
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", false, "Quiet copy")
+	cmd.Flags().BoolVar(&o.Delete, "delete", false, "Delete files not present in source")
+	cmd.Flags().BoolVar(&o.UseTar, "use-tar", false, "Use tar instead of rsync")
+	return cmd
+}
+
+func parseDestination(destination string) (string, string, error) {
+	parts := strings.SplitN(destination, ":", 2)
+	if len(parts) < 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
+		return "", "", fmt.Errorf("invalid destination %s: must be of the form PODNAME:DESTINATION_DIR", destination)
+	}
+	valid, msg := kvalidation.ValidatePodName(parts[0], false)
+	if !valid {
+		return "", "", fmt.Errorf("invalid pod name %s: %s", parts[0], msg)
+	}
+	return parts[0], parts[1], nil
+}
+
+// Complete verifies command line arguments and loads data from the command environment
+func (o *RsyncOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
+	switch n := len(args); {
+	case n == 0:
+		cmd.Help()
+		fallthrough
+	case n < 2:
+		return kcmdutil.UsageError(cmd, "SOURCE_DIR and POD:DESTINATION_DIR are required arguments")
+	case n > 2:
+		return kcmdutil.UsageError(cmd, "only SOURCE_DIR and POD:DESTINATION_DIR should be specified as arguments")
+	}
+
+	// Set main command arguments
+	o.Source = args[0]
+	o.Destination = args[1]
+
+	// Determine pod name
+	var err error
+	o.PodName, o.DestinationDir, err = parseDestination(o.Destination)
+	if err != nil {
+		return kcmdutil.UsageError(cmd, err.Error())
+	}
+
+	// Use tar if running on windows
+	// TODO: Figure out how to use rsync in windows so that I/O can be
+	// redirected from the openshift native command to the cygwin rsync command
+	if runtime.GOOS == "windows" {
```

Comment:
there should be some constant for "windows" ;-) just a nit I can live with this as well :)